src/main/java/org/broadinstitute/hellbender/engine/FeatureMergingWalker.java

```diff
@@ -0,0 +1,227 @@
+        }
+        return largeDict;
+    }
+
+    public static final class MergingIterator<F extends Feature> implements Iterator<PQEntry<F>> {
+        final SAMSequenceDictionary dict;
+        final PriorityQueue<PQEntry<F>> priorityQueue;
+
+        @SuppressWarnings("unchecked")
+        public MergingIterator( final SAMSequenceDictionary dict,
+                                final FeatureManager featureManager,
+                                final List<SimpleInterval> intervals ) {
+            final Set<FeatureInput<? extends Feature>> inputs = featureManager.getAllInputs();
+            this.dict = dict;
+            this.priorityQueue = new PriorityQueue<>(inputs.size());
+            for ( final FeatureInput<? extends Feature> input : inputs ) {
+                final Iterator<? extends Feature> iterator =
+                            featureManager.getFeatureIterator(input, intervals);
+                final Object header = featureManager.getHeader(input);
+                if ( iterator.hasNext() ) {
+                    addEntry((Iterator<F>)iterator, header);
+                }
+            }
+        }
+
+        @Override
+        public boolean hasNext() {
+            return !priorityQueue.isEmpty();
+        }
+
+        @Override
+        public PQEntry<F> next() {
+            final PQEntry<F> entry = priorityQueue.poll();
+            if ( entry == null ) {
+                throw new NoSuchElementException("iterator is exhausted");
+            }
+            final Iterator<F> iterator = entry.getIterator();
+            final Object header = entry.getHeader();
+            if ( iterator.hasNext() ) {
+                addEntry(iterator, header);
+            }
+            return entry;
+        }
+
+        private void addEntry( final Iterator<F> iterator, final Object header ) {
+            final F feature = iterator.next();
+            final int seqIdx = dict.getSequenceIndex(feature.getContig());
+            if ( seqIdx == -1 ) {
+                throw new UserException("dictionary has no entry for " + feature.getContig());
+            }
+            priorityQueue.add(new PQEntry<>(iterator, feature, seqIdx, header));
+        }
+    }
+
+    public static final class PQEntry<F extends Feature> implements Comparable<PQEntry<F>> {
+        private final Iterator<F> iterator;
+        private final F feature;
+        private final int seqIdx;
+        private final Object header;
+
+        public PQEntry( final Iterator<F> iterator, final F feature,
+                        final int seqIdx, final Object header ) {
+            this.iterator = iterator;
+            this.feature = feature;
+            this.seqIdx = seqIdx;
+            this.header = header;
+        }
+
+        public Iterator<F> getIterator() { return iterator; }
+        public F getFeature() { return feature; }
+        public Object getHeader() { return header; }
+
+        @Override
+        public int compareTo( PQEntry<F> entry ) {
```

Comment:
Could you use `IntervalUtils.compareLocatables()`?

Reply:
Sure.  If you want to slow it way down.
Less facetious explanation:  Every evidence object processed has to bubble up through the priority queue's heap and will undergo something like log2(nSamples) comparisons.  It's the inner-inner-innermost operation.  In this way of doing things, the expensive (relative to integer comparisons) hash lookup that translates a contig name into a dictionary index happens exactly once for each evidence object.  I haven't tested it, but I'm confident it's significantly faster.
I can race 'em if you're skeptical.

Reply:
Yeah.  That's probably true.  For small numbers of input files I/O time will probably dominate.